lib/Sema/SemaChecking.cpp

```diff
@@ -10666,22 +10669,32 @@ void Sema::DiagnoseCheckedDecl(const DeclaratorDecl *Decl) {
     Ty = Field->getType();
   }
 
-  bool checkedScopeError = false;
+  bool Res = true;
   int TypeKind = 0;
   if (TargetDecl) {
     // type is unchecked pointer/array type w/o bounds-safe interface
     // bounds safe interface is always checked type
-    QualType InterOpTy = GetCheckedCInteropType(TargetDecl);
-    if ((Ty->isUncheckedPointerType() || Ty->isUncheckedArrayType()) &&
-        InterOpTy.isNull()) {
-      checkedScopeError = true;
-      TypeKind = (Ty->isUncheckedPointerType() ? 0 : 1);
+
+    if (isa<PointerType>(Ty) || isa<ArrayType>(Ty)) {
+      // check if type is checked pointer type or it has interop type
+      QualType InterOpTy = GetCheckedCInteropType(TargetDecl);
+      if (!Ty->hasCheckedType() && InterOpTy.isNull()) {
+        Res = false;
+        TypeKind = (Ty->isUncheckedPointerType() ? 0 : 1);
+      }
     }
   }
-  if (checkedScopeError) {
-    Diag(TargetDecl->getLocStart(), diag::err_checked_scope_type) << DeclKind
+  if (!Res) {
```

Comment:
I think it is redundant to generate an error message at the declaration of a variable and the use of the variable, in the case where the variable was incorrectly declared to have an unchecked pointer type in a checked scope.  The additional error messages do not really give additional information to the programmer (they need to fix the original definition), and they definitely cause the programmer more work.

I'd suggest only generating an error message flagging a use of variable with unchecked pointer or array type in a checked scope if the variable was defined in a scope that was unchecked.